From 8beed8476c9a4d2c2082d254539e4e1050f86c19 Mon Sep 17 00:00:00 2001
From: kelson8 <kelson@kelsoncraft.net>
Date: Thu, 3 Sep 2026 17:59:00 -0400
Subject: [PATCH 074/161] feat: Make stat export read custom html template

* I had this working originally, I just didn't see where the output file got placed.

* Change colors in css for base template.
* Add library-versions.txt.

* Update make-x64.sh bash script, and c_cpp_properties.json.

* Update file-changes.md for guide.

* Move stat revc-stat-export html and css into gamefiles/ViceExtended/stats folder
* Add readme.txt into stats folder.
* Make CMenuManager::ExportStats output to 'ViceExtended/stats'.

* Update readme, and bump version number to 1.2.2, add alpha prefix back to this.
---
 .vscode/c_cpp_properties.json                 | 145 ++++++++++--------
 README.md                                     |  40 ++++-
 gamefiles/ViceExtended/stats/readme.txt       |  16 ++
 .../stats}/revc-stat-export-base.css          |  24 ++-
 .../stats}/revc-stat-export-base.html         |   0
 guides/file-changes.md                        |   8 +-
 library-versions.txt                          |   6 +
 make-x64.sh                                   |  19 ++-
 src/core/Frontend.cpp                         |  36 +++--
 src/extras/utils/defines.cpp                  |   2 +-
 10 files changed, 199 insertions(+), 97 deletions(-)
 create mode 100644 gamefiles/ViceExtended/stats/readme.txt
 rename gamefiles/{ => ViceExtended/stats}/revc-stat-export-base.css (64%)
 rename gamefiles/{ => ViceExtended/stats}/revc-stat-export-base.html (100%)
 create mode 100644 library-versions.txt

diff --git a/.vscode/c_cpp_properties.json b/.vscode/c_cpp_properties.json
index da8617e8..88822653 100644
--- a/.vscode/c_cpp_properties.json
+++ b/.vscode/c_cpp_properties.json
@@ -1,69 +1,80 @@
 {
-  "configurations": [
-    {
-      "name": "Mac",
-      "includePath": ["${default}"],
-      "defines": [],
-      "macFrameworkPath": [
-        "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks"
-      ],
-      "compilerPath": "/opt/local/bin/clang",
-      "compilerArgs": ["-g"],
-      "cStandard": "gnu11",
-      "cppStandard": "gnu++20",
-      "browse": {
-        "path": [
-          "/opt/local/include",
-          "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include"
-        ]
-      }
-    },
-    {
-      "name": "Linux",
-      "includePath": [
-        "${default}",
-        "lib/sol2/include"
+    "configurations": [
+        {
+            "name": "Mac",
+            "includePath": [
+                "${default}"
+            ],
+            "defines": [],
+            "macFrameworkPath": [
+                "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks"
+            ],
+            "compilerPath": "/opt/local/bin/clang",
+            "compilerArgs": [
+                "-g"
+            ],
+            "cStandard": "gnu11",
+            "cppStandard": "gnu++20",
+            "browse": {
+                "path": [
+                    "/opt/local/include",
+                    "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include"
+                ]
+            }
+        },
+        {
+            "name": "Linux",
+            "includePath": [
+                "${default}",
+                "lib/sol2/include",
+                "${workspaceFolder}/src/extras/test"
+            ],
+            "defines": [
+                "XDG_ROOT",
+                "RW_GL3"
+            ],
+            "compilerPath": "/usr/bin/gcc",
+            "compilerArgs": [
+                "-ggdb"
+            ],
+            "cStandard": "gnu11",
+            "cppStandard": "gnu++20"
+        },
+        {
+            "name": "devkitPro aarch64 (Nintendo Switch)",
+            "compilerPath": "${env:DEVKITPRO}/devkitA64/bin/aarch64-none-elf-g++",
+            "includePath": [
+                "${default}",
+                "${env:DEVKITPRO}/portlibs/switch/include",
+                "${env:DEVKITPRO}/libnx/include"
+            ],
+            "intelliSenseMode": "gcc-arm64",
+            "cStandard": "gnu11",
+            "cppStandard": "gnu++11",
+            "defines": [
+                "__SWITCH__",
+                "LIBRW",
+                "RW_GL3",
+                "AUDIO_OAL"
+            ]
+        },
+        {
+            "name": "Win32",
+            "includePath": [
+                "${workspaceFolder}/**",
+                "lib/sol2/include"
+            ],
+            "defines": [
+                "DEBUGMODE",
+                "RW_GL3"
+            ],
+            "windowsSdkVersion": "10.0.22621.0",
+            "compilerPath": "cl.exe",
+            "cStandard": "c17",
+            "cppStandard": "c++20",
+            "intelliSenseMode": "windows-msvc-x64",
+            "configurationProvider": "ms-vscode.makefile-tools"
+        }
     ],
-      "defines": [
-        "XDG_ROOT",
-        "RW_GL3"
-      ],
-      "compilerPath": "/usr/bin/gcc",
-      "compilerArgs": ["-ggdb"],
-      "cStandard": "gnu11",
-      "cppStandard": "gnu++20"
-    },
-    {
-      "name": "devkitPro aarch64 (Nintendo Switch)",
-      "compilerPath": "${env:DEVKITPRO}/devkitA64/bin/aarch64-none-elf-g++",
-      "includePath": [
-        "${default}",
-        "${env:DEVKITPRO}/portlibs/switch/include",
-        "${env:DEVKITPRO}/libnx/include"
-      ],
-      "intelliSenseMode": "gcc-arm64",
-      "cStandard": "gnu11",
-      "cppStandard": "gnu++11",
-      "defines": ["__SWITCH__", "LIBRW", "RW_GL3", "AUDIO_OAL"]
-    },
-
-    {
-      "name": "Win32",
-      "includePath": [
-        "${workspaceFolder}/**",
-        "lib/sol2/include"
-      ],
-      "defines": [
-        "DEBUGMODE",
-        "RW_GL3"
-      ],
-      "windowsSdkVersion": "10.0.22621.0",
-      "compilerPath": "cl.exe",
-      "cStandard": "c17",
-      "cppStandard": "c++20",
-      "intelliSenseMode": "windows-msvc-x64",
-      "configurationProvider": "ms-vscode.makefile-tools"
-    }
-  ],
-  "version": 4
-}
+    "version": 4
+}
\ No newline at end of file
diff --git a/README.md b/README.md
index c4d1e07d..cf7b8258 100644
--- a/README.md
+++ b/README.md
@@ -28,15 +28,32 @@ Added glfw-3.4.bin.WIN64 to the vendor folder for OpenGL building to work.
 
 This seems to fix the glitchy audio that I was having before, I think it was possibly my lua setup that was causing it.
 
+I have disabled D3D9 and DirectX builds for this ReVC build, I didn't want to support both for ImGui and OpenGL works well on Linux, I'm not too worried about DirectX support for this ReVC.
+
+By default, this uses C++20 instead of C++11, but it runs fine with C++20 setup.
+
 ## ImGui info
 This project has an ImGui mod menu, which can be opened with `F8` on the keyboard, or `RB` + `A` on an Xbox controller.
 
 I got this building and running on Windows and Linux.
 Tested ImGui on my Steam deck and now it seems to work, I accidentally disabled the ImGui init for other platforms in glfw.cpp.
 
+You can toggle ImGui in the `reVC.ini`, if `ImGuiMenuToggle` is set to 0 it is off, and 1 is on.
+
 Librw is now added in as a submodule within the opengl-imgui branch
 * https://git.internal.kelsoncraft.net/kelson8/librw/src/branch/opengl-imgui/
 
+## Lua Info
+
+This ReVC project has a lua loader that loads custom lua scripts.
+
+Here is a list of info for the lua scripts
+[New files and additions](guides/file-changes.md#new-files-and-additions)
+
+There are some example scripts in `gamefiles/ViceExtended/lua_scripts`
+
+You can toggle Lua in the `reVC.ini`, if `LuaTest` is set to 0 it is off, and 1 is on.
+
 ## Docker support
 This branch now has support for building with Docker on Linux, tested on WSL with Ubuntu 24.04 and a Debian 13 server.
 
@@ -49,6 +66,16 @@ Currently, this doesn't give you a way to view the built files in a volume outsi
 To view the files in the container (Specify version instead of writing version):
 * docker run --rm -it revc-dev:version /bin/bash
 
+## TODO
+
+Look into setting up GDB with VSCode for reVC
+* https://medium.com/@ogawa.henrique/gdb-debugging-with-vscode-754758837603
+
+Setup CMake for this reVC build with ImGui, Lua and other needed stuff for my build.
+
+Setup LuaBridge3 and namespaces with Lua, I mostly got this working but I need to figure out how
+ to access values from the Lua code, such as logging a message.
+
 ## Installation
 
 - reVC requires game assets to work, so you **must** own [a copy of GTA Vice City](https://store.steampowered.com/app/12110/Grand_Theft_Auto_Vice_City/).
@@ -153,7 +180,7 @@ We accept only these kinds of PRs;
 
 We have a [Coding Style](https://github.com/GTAmodding/re3/blob/master/CODING_STYLE.md) document that isn't followed or enforced very well.
 
-Do not use features from C++11 or later.
+Features from C++17 and later are now used in this ReVC build.
 
 
 ## History
@@ -193,6 +220,17 @@ I added d3dx9 into vendor/d3d9 from these GitHub links
 GLFW3 was obtained from the official GLFW website here
 * https://www.glfw.org/
 
+This project uses Lua, which is added into `lib/lua`, and ImGui which is added into the lib folder as a submodule.
+
+Also, this project uses a custom modified build of librw to add support for the ImGui mod menu in the rendering code, since I only have this working for OpenGL and have disabled D3D9 support it requires the opengl branch from my Git repo here [librw Gitea](https://git.internal.kelsoncraft.net/kelson8/librw/src/branch/opengl-imgui/).
+
+
+## Credits
+Credit goes to user-grinch on GitHub for most of the files under the `gamefiles/ViceExtended/toml` folder from the [Cheat Menu](https://github.com/user-grinch/Cheat-Menu/releases/tag/3.52) mod, and also for the mod menu styles along with the custom font this mod menu uses.
+
+Credit goes to x87 on GitHub for some of the extra code that I am using which comes from the [gta-extended](https://github.com/x87/gta-extended) project, I may try to implement the swimming feature from this code in the future .
+
+
 ## License
 
 We don't feel like we're in a position to give this code a license.\
diff --git a/gamefiles/ViceExtended/stats/readme.txt b/gamefiles/ViceExtended/stats/readme.txt
new file mode 100644
index 00000000..f10e836a
--- /dev/null
+++ b/gamefiles/ViceExtended/stats/readme.txt
@@ -0,0 +1,16 @@
+-------------
+Folder Info
+
+This is where the stats.html and stats.txt get output when saved from the game.
+
+The revc-stat-export-base.html file is required for the game to be able to export the stats to the file.
+And the revc-stat-export-base.css file contains the custom CSS for the stats.html file.
+
+You can easily change the text color by setting some of the values in the CSS.
+It is recommended to leave the html file alone, it is setup a specific way for the games code to read it.
+
+-------------
+Technical info
+
+I hated how the game was using the C++ code for the output html, which made this a pain to debug and change.
+So I made it read from a base template which will be added into from CMenuManager::ExportStats() in Frontend.cpp.
diff --git a/gamefiles/revc-stat-export-base.css b/gamefiles/ViceExtended/stats/revc-stat-export-base.css
similarity index 64%
rename from gamefiles/revc-stat-export-base.css
rename to gamefiles/ViceExtended/stats/revc-stat-export-base.css
index a7beb757..8849b4b3 100644
--- a/gamefiles/revc-stat-export-base.css
+++ b/gamefiles/ViceExtended/stats/revc-stat-export-base.css
@@ -1,7 +1,11 @@
 
 
 body {
-    background-color: #ff00cc;
+    /* Original pink color */
+/*     background-color: #ff00cc; */
+    /* Black */
+    background-color: #000000;
+
     margin: 10px;
     font-family: Arial, Helvetica, sans-serif;
 }
@@ -29,15 +33,29 @@ body {
 
 .rating {
     height: 20px;
-    color: #ff00cc;
+    /* Original pink color */
+/*     background-color: #ff00cc; */
+    /* Black */
+    background-color: #000000;
     background-color: #ffffff;
 }
 
 .stat-row td {
     height: 22px;
     padding: 5px;
+
+
+    /* Original pink color */
     background-color: #ffccff;
-    color: #ff00cc;
+    /* Black */
+/*     background-color: #000000; */
+
+
+
+    /* Original pink color */
+/*     color: #ff00cc; */
+    /* Black */
+    color: #000000;
 }
 
 .stat-name {
diff --git a/gamefiles/revc-stat-export-base.html b/gamefiles/ViceExtended/stats/revc-stat-export-base.html
similarity index 100%
rename from gamefiles/revc-stat-export-base.html
rename to gamefiles/ViceExtended/stats/revc-stat-export-base.html
diff --git a/guides/file-changes.md b/guides/file-changes.md
index 3d4547cd..cd6eaa82 100644
--- a/guides/file-changes.md
+++ b/guides/file-changes.md
@@ -35,16 +35,14 @@ Lua info for devs:
 | kcnet-busted.lua  | Running under `WBSTATE_BUSTED` in `GameLogic::Update` |
 | kcnet-keybind-events.lua  | This runs under `LuaSetup::LuaPlayingGameLoop` if the F9 key is pressed, I need to make this configurable later. |
 | kcnet-wasted.lua | Running under `WBSTATE_WASTED` in `GameLogic::Update`. |
-| test.lua | I'm not sure what this is doing, it seems to only run when the game starts up under `LuaTest::InitLua` |
-
-* test.lua - This is running under 'WBSTATE_WASTED' and 'WBSTATE_BUSTED' switch statements in GameLogic.cpp, and in the 'GS_PLAYING_GAME' switch statement within win.cpp.
+| test.lua | This is running under 'WBSTATE_WASTED' and 'WBSTATE_BUSTED' switch statements in GameLogic.cpp, and in the 'GS_PLAYING_GAME' switch statement within glfw.cpp. |
 
 'ViceExtended' folder {My additions} (TODO Move these into 'ViceExtended/kcnet' folder or something):
 
 My items:
 * ImGui-Controls.ini - This will contain a config for xbox controller keybinds once I setup the config for ImGui with it, I don't have the config for Xbox keybinds working in OpenGL yet.
-* vehicles.ini - List of vehicles from the Cheat Menu of user-grinch on github, used by my ImGui menu.
-* warps.ini - List of warps from the Cheat Menu of user-grinch on github, used by my ImGui menu.
+* vehicles.ini - List of vehicles from the [Cheat Menu](https://github.com/user-grinch/Cheat-Menu/releases/tag/3.52) of user-grinch on github, used by my ImGui menu.
+* warps.ini - List of warps from the [Cheat Menu](https://github.com/user-grinch/Cheat-Menu/releases/tag/3.52) of user-grinch on github, used by my ImGui menu.
 
 'ViceExtended' folder:
 * gamecontrollerdb.txt - I think this is from the ReVC project originally, just moved into here.
diff --git a/library-versions.txt b/library-versions.txt
new file mode 100644
index 00000000..08a239f6
--- /dev/null
+++ b/library-versions.txt
@@ -0,0 +1,6 @@
+This is a list of versions for the libraries that I am using
+
+ImGui: v1.92.9 (https://github.com/ocornut/imgui/commit/11acd9f5cb469890c3915e36f8b8ea702add3c95)
+LuaBridge3 (Not currently in use): 3.0-rc12 (https://github.com/kunitoki/LuaBridge3/releases/tag/3.0-rc12)
+Lua: 5.3 (I think this is correct, found in Makefile as 5_3)
+Librw: Custom modified build for ImGui and OpenGL (https://github.com/kelson8/librw/tree/opengl-imgui)
diff --git a/make-x64.sh b/make-x64.sh
index b703a3be..4700bdd2 100755
--- a/make-x64.sh
+++ b/make-x64.sh
@@ -4,15 +4,18 @@
 
 BUILD_FOLDER=build
 DEBUG_BUILD=debug_linux-amd64-librw_gl3_glfw-oal
-RELEASE_BUILD=release_linux-amd64-librw_gl3_glfw-oal
+# RELEASE_BUILD=release_linux-amd64-librw_gl3_glfw-oal
 
-if [ ! -d build ]; then
+# if [ ! -d build ]; then
+if [ ! -d "$BUILD_FOLDER" ]; then
     echo "Build folder not found, running premake."
-    ./premake5Linux --with-librw --imgui-opengl --no-git-hash gmake2
+    # Changed with Dockerfile params, might fix the SHA256 checksum between Docker and without it.
+    # Well this didn't fix it, it still builds but the SHA256sum when run without Docker is different then
+    # if ran with Docker for some reason.
+    ./premake5Linux --no-git-hash=1 --with-librw --imgui-opengl gmake2
 fi
 
-cd build
-# make config=debug_linux-amd64-librw_gl3_glfw-oal -j$(nproc)
-make config=$DEBUG_BUILD -j$(nproc)
-make config=$RELEASE_BUILD -j$(nproc)
-# make config=release_linux-amd64-librw_gl3_glfw-oal -j$(nproc)
+cd "$BUILD_FOLDER" || exit
+make config=$DEBUG_BUILD -j"$(nproc)"
+#make config=$RELEASE_BUILD -j"$(nproc)"
+
diff --git a/src/core/Frontend.cpp b/src/core/Frontend.cpp
index 1a904316..0d29235c 100644
--- a/src/core/Frontend.cpp
+++ b/src/core/Frontend.cpp
@@ -3826,7 +3826,14 @@ void
 CMenuManager::ExportStats()
 {
 	char date[10];
+
+#ifdef CUSTOM_CODE
+	// Switch folder to ViceExtended/stats
+	CFileMgr::SetDir("ViceExtended/stats");
+#else
 	CFileMgr::SetDirMyDocuments();
+#endif // CUSTOM_CODE - Move ExportStats output to ViceExtended/stats folder.
+
 	_strdate(date);
 	wchar *lastMission = TheText.Get(CStats::LastMissionPassedName[0] == '\0' ? "ITBEG" : CStats::LastMissionPassedName);
 	FILE *txtFile = fopen("stats.txt", "w");
@@ -3864,14 +3871,20 @@ CMenuManager::ExportStats()
 	fclose(txtFile);
 
 // Disable this for now.
-// TODO Fix this to work.
-#undef CUSTOM_CODE
 #ifdef CUSTOM_CODE
-	// TODO Implement this, read these values from the html template.
-	// So far, this is incomplete.
+	// Places the generated stats.html file into the ViceExtended folder.
+	// This uses a HTML template that I have setup. 
+	// It can easily be modified without rebuilding the code.
+
+	// There is no reason to have the HTML hardcoded into here.
+	// But, this can be disabled by placing '#undef CUSTOM_CODE' above this function.
 
 	// This should set the folder to ViceExtended for this template.
-	CFileMgr::SetDir("ViceExtended");
+	// The HTML stats file gets output to 'ViceExtended/stats/stats.html' in the ReVC folder.
+	
+	// TODO Remove this later.
+	// I moved this near the top of the function to also output the txt file to this folder.
+	// CFileMgr::SetDir("ViceExtended/stats/");
 
 	int statLines = CStats::ConstructStatLine(99999);
 
@@ -3883,7 +3896,7 @@ CMenuManager::ExportStats()
 
 	if(html.empty())
 	{
-		TRACE("Html template is empty or could not be read.");
+		TRACE("[KCNet-Warning]: Html template is empty or could not be read.");
 		CFileMgr::SetDir("");
 		return;
 	}
@@ -3923,15 +3936,14 @@ CMenuManager::ExportStats()
 	for (int k = 0; k < statLines; ++k) {
     	CStats::ConstructStatLine(k);
 
-    	char* statKey = UnicodeToAscii(gUString);
-    	char* statValue = UnicodeToAscii(gUString2);
-
-    	std::string key = statKey ? statKey : "";
-    	std::string value = statValue ? statValue : "";
+		const char* keyText = UnicodeToAscii(gUString);
+		std::string key = keyText ? keyText : "";
+		const char* valueText = UnicodeToAscii(gUString2);
+		std::string value = valueText ? valueText : "";
 
     	for (char& character : value) {
         	if (character == '_')
-            	character = '\xBA';
+            	character = '\xBA'; // This is degree symbol, but my editors keeps messing up with it so I wrote hex representation
     	}
 
     	statRows +=
diff --git a/src/extras/utils/defines.cpp b/src/extras/utils/defines.cpp
index 00f886cf..24215766 100644
--- a/src/extras/utils/defines.cpp
+++ b/src/extras/utils/defines.cpp
@@ -86,7 +86,7 @@ std::string Defines::modificationAuthor = "kelson8";
 // TODO Make this grab from my latest internal gitea releases or something.
 // std::string Defines::programVersion = "1.1";
 #if defined(GTAVC)
-std::string Defines::programVersion = "1.2.1";
+std::string Defines::programVersion = "1.2.2a";
 #elif defined(GTA3)
 std::string Defines::programVersion = "1.2a - Pre Release";
 #elif defined(GTALCS)
-- 
2.43.0

